walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/servlets/FindMeetingQuery.java

```diff
@@ -0,0 +1,152 @@
+{
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request)
+    {
+        long meetingDuration = request.getDuration();
+        Collection<String> meetingAttendees = request.getAttendees();
+        Collection<String> optionalAttendees = request.getOptionalAttendees();
+        Collection<TimeRange> availableTime = new ArrayList<TimeRange>(); 
+        Collection<TimeRange> removeList = new ArrayList<TimeRange>();
+        Collection<TimeRange> addList = new ArrayList<TimeRange>();
+        Collection<TimeRange> finalTimeList = Arrays.asList();
+        if (meetingDuration > TimeRange.WHOLE_DAY.duration())
+        {
+            finalTimeList = Arrays.asList();
+        }
+        //If there are no meeting attendees or no events to look at 
+        else if ((meetingAttendees.isEmpty()&&optionalAttendees.isEmpty()) || events.isEmpty())
+        {
+            finalTimeList = Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+        else
+        {
+            for(Event event : events)
+            {
+                TimeRange whenEvent = event.getWhen();
+                Set<String> eventAttendees = event.getAttendees();
+                //Check if event attendees are invited to meeting OR there are no mandotory attendees/only optionals
+                if (checkAttendees(eventAttendees,meetingAttendees)||
+                    (meetingAttendees.isEmpty() && !optionalAttendees.isEmpty()))
+                {
+                    if (availableTime.isEmpty())
+                    {
+                        TimeRange frontTime1= TimeRange.fromStartEnd(TimeRange.START_OF_DAY,whenEvent.start(),false);
+                        TimeRange backTime1 = TimeRange.fromStartEnd(whenEvent.end(),TimeRange.END_OF_DAY,true);
+                        addFrontAndBackTimesToList(meetingDuration, frontTime1, backTime1, availableTime);
+                    }
+                    else
+                    {
+                        for (TimeRange openTime : availableTime)
+                        {
+                            //check if times overlap to produce the right time slots around the current event
+                            if(openTime.overlaps(whenEvent))
+                            {
+                                if(openTime.contains(whenEvent))
+                                {
+                                    removeList.add(openTime);
+                                    TimeRange frontTime2 = TimeRange.fromStartEnd(openTime.start(),whenEvent.start(),false);
+                                    TimeRange backTime2 = TimeRange.fromStartEnd(whenEvent.end(),TimeRange.END_OF_DAY,true);
+                                    addFrontAndBackTimesToList(meetingDuration, frontTime2, backTime2,addList);
+                                }
+                                else if (whenEvent.start() < openTime.start())
+                                {
+                                    removeList.add(openTime);
```

Comment:
The logic in this method is very nested and it is confusing to understand which cases we're covering in which branches. Can you break out parts of the logic into well-named subroutines so that the structure of the algorithm becomes more visible?